### PR TITLE
MOTECH-2202: changed openmrs person model to support all address fields

### PR DESCRIPTION
**Pull request by sebastianceronik**
Added Fields to Person/PreferredAdress :
address2, cityVillage, stateProvince, country, postalCode, countryDistrict, address3, address4, address5, address6, startDate, endDate, latitude, longitude, and preferred.

setting up preferred adress to true while creating a new patient in openmrs.

### CODE REVIEW

File: openmrs-19/src/main/java/org/motechproject/openmrs19/util/ConverterUtils.java [hunk 1]
```diff
@@ -429,7 +444,7 @@ public static Patient toPatient(OpenMRSPatient patient, OpenMRSPerson savedPerso
         Person person = new Person();
         person.setUuid(savedPerson.getPersonId());
         converted.setPerson(person);
-
+        converted.getPerson().getPreferredAddress().setPreferred("true");
```
Comment (kmadej) [inline openmrs-19/src/main/java/org/motechproject/openmrs19/util/ConverterUtils.java]: can you explain why you add this ?

btw. you can setPreferred before setting person in converted

File: openmrs-19/src/main/java/org/motechproject/openmrs19/resource/model/Person.java [hunk 2]
```diff
@@ -288,8 +288,54 @@ public boolean equals(Object obj) {
...
+
+        @Expose
+        private String endDate;
+
+        @Expose
+        private String latitude;
+
+        @Expose
+        private String longtitude;
+
```
Comment (kmadej) [inline openmrs-19/src/main/java/org/motechproject/openmrs19/resource/model/Person.java]: remove line

File: openmrs-19/src/main/java/org/motechproject/openmrs19/resource/model/Person.java [hunk 3]
```diff
@@ -306,6 +360,121 @@ public void setAddress1(String address1) {
...
+        public String getLongtitude() {
+            return longtitude;
+        }
+
+        public void setLongtitude(String longtitude) {
+            this.longtitude = longtitude;
+        }
+
+
+
```
Comment (kmadej) [inline openmrs-19/src/main/java/org/motechproject/openmrs19/resource/model/Person.java]: remove lines

File: openmrs-19/src/main/java/org/motechproject/openmrs19/resource/model/Person.java [hunk 4]
```diff
@@ -306,6 +360,121 @@ public void setAddress1(String address1) {
...
+
         @Override
         public int hashCode() {
```
Comment (kmadej) [inline openmrs-19/src/main/java/org/motechproject/openmrs19/resource/model/Person.java:477]: do we really need all this fields ?  if yes please add fields to hashcode and equals.

File: openmrs-19/src/main/java/org/motechproject/openmrs19/util/ConverterUtils.java [hunk 5]
```diff
@@ -66,7 +66,22 @@ public static OpenMRSPerson toOpenMRSPerson(Person person) {
         converted.setGender(person.getGender());
 
         if (person.getPreferredAddress() != null) {
-            converted.setAddress(person.getPreferredAddress().getAddress1());
+            converted.setAddress(
```
Comment (kmadej) [inline openmrs-19/src/main/java/org/motechproject/openmrs19/util/ConverterUtils.java]: 1. use StringBuilder
2. move converting method to Person model
3. add unit test

File: openmrs-19/src/main/java/org/motechproject/openmrs19/resource/model/Person.java [hunk 6]
```diff
@@ -288,8 +288,54 @@ public boolean equals(Object obj) {
...
+        private String startDate;
+
+        @Expose
+        private String endDate;
+
+        @Expose
+        private String latitude;
+
+        @Expose
+        private String longtitude;
```
Comment (llavoie) [inline openmrs-19/src/main/java/org/motechproject/openmrs19/resource/model/Person.java]: Longitude is misspelled.

File: openmrs-19/src/main/java/org/motechproject/openmrs19/resource/model/Person.java [hunk 7]
```diff
@@ -306,12 +359,137 @@ public void setAddress1(String address1) {
...
+
+        public String getLongitude() {
+            return longitude;
+        }
+
+        public void setLongitude(String longitude) {
+            this.longitude = longitude;
+        }
+
+        public String getFullAdressString() {
```
Comment (kmadej) [inline openmrs-19/src/main/java/org/motechproject/openmrs19/resource/model/Person.java]: address

File: openmrs-19/src/main/java/org/motechproject/openmrs19/util/ConverterUtils.java [hunk 8]
```diff
@@ -125,9 +125,7 @@ public static Person toPerson(OpenMRSPerson person, boolean includeNames) {
         }
         converted.setBirthdateEstimated((Boolean) ObjectUtils.defaultIfNull(person.getBirthDateEstimated(), false));
         converted.setDead((Boolean) ObjectUtils.defaultIfNull(person.getDead(), false));
-        if (person.getCauseOfDeath() != null) {
-            converted.setCauseOfDeath(toConcept(person.getCauseOfDeath()));
-        }
+        converted.setCauseOfDeath(toConcept(person.getCauseOfDeath()));
```
Comment (kmadej) [inline openmrs-19/src/main/java/org/motechproject/openmrs19/util/ConverterUtils.java]: do not remove this

File: openmrs-19/src/main/java/org/motechproject/openmrs19/util/ConverterUtils.java [hunk 9]
```diff
@@ -428,6 +426,7 @@ public static Patient toPatient(OpenMRSPatient patient, OpenMRSPerson savedPerso
         Patient converted = new Patient();
         Person person = new Person();
         person.setUuid(savedPerson.getPersonId());
+        person.getPreferredAddress().setPreferred("true");
```
Comment (kmadej) [inline openmrs-19/src/main/java/org/motechproject/openmrs19/util/ConverterUtils.java]: I'm not sure about preferred field. In OpenMRS server Person has fields addresses and preferredAddress. PreferredAddress is one of object in addresses list with flag preferred set to true. In our Motech implementation we support only preferredaddress. 
1. I noticed that you set this flag in method toPatient. Here you only set person id in patient object. You should do this in toPerson method. 
2. It should be boolean field not string
3. Please check if this field is required. I suppose that OpenMRS module stores address as preferred and this flag is set automatically

**Comment by kmadej** [comment]
@sebastianceronik please add a unit test and integration test in another PR

**Comment by sebastianceronik** [comment]
Removed setting up preferred flag to true since it was unnecessary to set it up.
It is set as true on openMRS by default while creating patient

**Pull request merged**